### PR TITLE
SFB-218: logic begrotingstabel total field

### PR DESCRIPTION
**Pull request by JonasVanHoof**
## ID
 [SFB-218](https://binnenland.atlassian.net/browse/SFB-218)

 ## Description

We want to calculate the sum of values in a column. These values will be shown in a total field. There will be added more fields depending on how many targets you add to the begrotingsTotal field. These values will be shown in the form preview. When these values do not end up being equal to each other we will show a message that it is not valid.

 ## Type of change

 - [ ] Bug fix
 - [x] New feature
 - [ ] Breaking change
 - [ ] Other

 ## How to test

1. Create a new form from template. Choose `Begrotingstabel`
2. Start adding values to the `Inkomsten` and `Uitgaven` tables
3. If the result values do not match each other a error message is shown

 ## Links to other PR's

 - https://github.com/lblod/frontend-form-builder/pull/144

 ## Attachements

**V1**

image

**V2**

image

**V3**

image

### CODE REVIEW

File: app/components/rdf-form-fields/begrotingstabel-total-field.js [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    }).format(amount);
+  }
+
+  calculateTotals = restartableTask(async () => {
+    const collections = this.store
+      .match(this.field.uri, XSD('target'), undefined, this.graphs.formGraph)
+      .map((st) => st.object);
+
+    if (collections.length !== 2) {
+      this.collectionErrorMessage = `Tabel heeft te weing informatie. Voeg twee xsd:targets toe. bv. xsd:target ( ext:Expense ext:amount ), ( ext:Income ext:amount );`;
```
Comment (JonasVanHoof) [inline app/components/rdf-form-fields/begrotingstabel-total-field.js:45]: Should we add a translation file for just the `BegrotingstabelTotal` component?

File: app/components/rdf-form-fields/begrotingstabel-total-field.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    this.setWarningMessage();
+  });
+
+  getIndexOfTotal(collectionId) {
+    return this.totals.findIndex((item) => item.id == collectionId);
+  }
+
+  isValidIndex(index) {
+    return index == 0 || index !== -1;
+  }
```
Comment (JonasVanHoof) [inline app/components/rdf-form-fields/begrotingstabel-total-field.js:94]: Generic utils method? is used in two places now

File: app/components/rdf-form-fields/begrotingstabel-total-field.js [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+    if (
+      !this.totals.every(
+        (item) => item.calculationResult == this.totals[0].calculationResult
+      )
+    ) {
+      this.comparisonWarningMessage = `De resultaten zijn niet gelijk`;
+
+      return;
+    }
```
Comment (JonasVanHoof) [inline app/components/rdf-form-fields/begrotingstabel-total-field.js:116]: Could also check if value on index 0 and 1 are the same

File: app/components/rdf-form-fields/begrotingstabel-total-field.js (hunk 1 above)
Comment (JonasVanHoof) [reply app/components/rdf-form-fields/begrotingstabel-total-field.js:45]: These translations can than be easely added to the project that will be using this component.

**Review by ZijneDwaasheid — APPROVED** [review]

File: public/forms/templates/begrotingstabel.ttl [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        nodes:f0d14635-879d-44ff-ae9b-2c34ed278f8a;
+    form:partOf
+        nodes:4ce76f84-e71d-4d2b-8733-7fb02cbb12f7,
+        nodes:a4503b20-7c8f-497c-abac-9650691cec75,
+        nodes:f9dee3f6-e782-417f-ad89-c54e603a6a04 .
+
+ext:expenseGenerator
+    a form:Generator;
+    form:dataGenerator form:addMuUuid;
+    form:prototype [ form:shape [ a ext:Expense; ext:amount 0.0 ] ].
```
[the remark this answers is not in the document]
Comment (JonasVanHoof) [reply public/forms/templates/begrotingstabel.ttl:132]: It depends, when clicking before the zero and typing your number the zero dissapears but if you click after the zero it means you want to make the number bigger.

File: app/components/rdf-form-fields/begrotingstabel-total-field.js (hunk 2 above)
Comment (JonasVanHoof) [reply app/components/rdf-form-fields/begrotingstabel-total-field.js:94]: No as this is a component that needs to be copied to other projects aswell if they want to use it there

**Pull request merged**